fdk/http_stream.py

```diff
@@ -0,0 +1,110 @@
+
+import ujson
+
+from aiohttp import web
+from xml.etree import ElementTree
+
+from fdk import constants
+from fdk import log
+from fdk import runner
+from fdk import response as rtypes
+
+
+def serialize_response_data(data, content_type):
+    log.log("in serialize_response_data")
+    if data:
+        if "application/json" in content_type:
+            return bytes(ujson.dumps(data), "utf8")
+        if "text/plain" in content_type:
+            return bytes(str(data), "utf8")
+        if "application/xml" in content_type:
+            # returns a bytearray
+            if isinstance(data, str):
+                return bytes(data, "utf8")
+            return ElementTree.tostring(data, encoding='utf8', method='xml')
+        if "application/octet-stream" in content_type:
+            return data
+    return
+
+
+def handle(handle_func):
+    async def pure_handler(request):
+        log.log("in pure_handler")
+        data = None
+        if (request.has_body and
+                request.content_length is not None and
+                request.content_length > 0):
+            log.log("request comes with data")
+            data = await request.content.read(request.content_length)
+
+        response = await runner.handle_request(
+            handle_func, None, constants.HTTPSTREAM,
+            request=request, data=data)
+        log.log("request execution completed")
+        headers = (response.headers()
+                   if isinstance(response, rtypes.RawResponse)
+                   else response.headers)
+        response_content_type = headers.get(
+            "content-type", "application/json"
+        )
+        headers.set("content-type", response_content_type)
+        kwargs = {
+            "status": response.status(),
+            "headers": headers.http_raw()
+        }
+
+        sdata = serialize_response_data(
+            response.body(), response_content_type)
+
+        if response.status() >= 500:
+            kwargs.update(reason=sdata)
+        else:
+            kwargs.update(body=sdata)
+
+        log.log("sending response back")
+        try:
+            resp = web.Response(**kwargs)
+        except (Exception, BaseException) as ex:
+            resp = web.Response(
+                text=str(ex), reason=str(ex),
+                status=500, content_type="text/plain")
+
+        return resp
+
+    return pure_handler
+
+
+def setup_unix_server(handle_func, loop=None):
+    log.log("in setup_unix_server")
+    app = web.Application(loop=loop)
+
+    for m in [app.router.add_get,
+              app.router.add_post,
+              app.router.add_put,
+              app.router.add_patch,
+              app.router.add_delete,
+              app.router.add_head]:
+        m('/{tail:.*}', handle(handle_func))
+
+    return app
+
+
+def start(handle_func, uds, loop=None):
+    log.log("in htt_stream.start")
+    app = setup_unix_server(handle_func, loop=loop)
+    web.run_app(app, path=uds[len("unix:"):],
```

Comment:
Is the socket here guaranteed to : 

* be listening when it is created
* be globally writable when it is created 

if that's not the case you might see race conditions with Fn - in Java and go we had to create the socket, chmod it and then symlink it after creation to guarantee the above.